### PR TITLE
[lineage-20] Merge CIP 4.4.y-st updates

### PR DESCRIPTION
**Pull request by derfelot**
Up to current tag (4.4-st25)

Still needs testing.

### CODE REVIEW

**Comment by derfelot** [comment]
Now up to 4.4-st37

To be merged once tested

**Comment by derfelot** [comment]
Turns out that https://github.com/derfelot/android_kernel_sony_msm8998/commit/5e7f0d792dd34d916706d14dba0ec3530e77f310 is making it not boot. Either I messed something up fixing the conflicts, or it actually needs more love (we are deviated quite a bit from Linux upstream here). I don't think this is really worth the effort in the end, as it doesn't seem to fix anything crucial for us.

Will revert it for now and maybe revisit later.

File: drivers/mmc/core/sdio_bus.c [hunk 1]
```diff
@@ -273,6 +273,8 @@ static void sdio_release_func(struct device *dev)
 	 * cis tables for this func
 	 */
 	if (!func->card->host->embedded_sdio_data.funcs)
+#else
```
Comment (Flamefire) [inline drivers/mmc/core/sdio_bus.c:276]: I'm not sure about this `#else`: To me it looks like the `sdio_free_func_cis` function should not be called for embedded SDIO **or** non-standard SDIO. We don't seem to use `CONFIG_MMC_EMBEDDED_SDIO` so it shouldn't matter in practice but I think this `#else` needs to be the `#endif` from https://github.com/whatawurst/android_kernel_sony_msm8998/pull/82/files#diff-0b4db9467c20a99828df56af64d17a2fa58aa1ed240ef133078480a15283437eR278

**Comment by Flamefire** [comment]
> Turns out that [derfelot@5e7f0d7](https://github.com/derfelot/android_kernel_sony_msm8998/commit/5e7f0d792dd34d916706d14dba0ec3530e77f310) is making it not boot. Either I messed something up fixing the conflicts, or it actually needs more love (we are deviated quite a bit from Linux upstream here). I don't think this is really worth the effort in the end, as it doesn't seem to fix anything crucial for us.
> 
> Will revert it for now and maybe revisit later.

I [merged the CIP branch just now into my LOS 17.1 build](https://github.com/Flamefire/android_kernel_sony_msm8998/pull/42) and were especially careful with the commit you mentioned. The 17.1 build boots successfully but the merge was not trivial. Especially the IRQ lock/unlock could be the culprit: https://github.com/whatawurst/android_kernel_sony_msm8998/pull/82/commits/54e69082291a3e34538697693beceeef81dcc18b#diff-6db4e3c76a0a028eee97050a4e65b98db4719401bf5886925e23403264fb8a65R2088

It makes sense to me to put the `out` label before `spin_unlock_irqrestore`, not after. Following the controlflow this makes sure that it is unlocked in all cases I could see and I haven't found a way for this to be reached when it is not yet locked or already unlocked.

**Pull request merged**